076-moving-data-between-two-jbod-disks-using-cruise-control.md

```diff
@@ -0,0 +1,87 @@
+
+This proposal is about integrating the [`remove_disks`](https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/resources/yaml/endpoints/removeDisks.yaml) endpoint from Cruise Control into Strimzi cluster operator. 
+This endpoint will allow us to move the data between two JBOD disks. 
+
+## Current situation
+
+Currently, we get a multiple requests from community users to add the ability for moving all Kafka logs between two disks on the JBOD storage array. This feature can be useful in following scenarios:
+- The current disk is too small and the user wants to use a bigger one, or vice versa.
+- When we want to use a different Storage Class with different parameters or different storage types.
+- In case of disk removal to reduce the total storage.
+
+For now, we can do this using the Kafka CLI tool i.e. `kafka-reassign-partitions.sh` tool, but it takes a lot of manual steps which is time-consuming and not so user-friendly.
+
+## Motivation
+
+We should introduce the logic to Strimzi to leverage Cruise Control integration and make it possible to move the data between two JBOD disks.
+This feature will also allow us to remove the disks without the loss of data.
+
+## Proposal
+
+Cruise Control provides `remove_disks` HTTP REST endpoint to move replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+This endpoint triggers a rebalancing operation by moving replicas in a size-based manner to the remaining disks, from the largest to the smallest, while checking the following constraint:
+```sh
+1 - (remainingUsageAfterRemoval / remainingCapacity) > errorMargin
+```
+where:
+```sh
+remainingUsageAfterRemoval = current usage for remaining disks + additional usage from removed disks
+remainingCapacity = sum of capacities of the remaining disks
+errorMargin = configurable property (default 0.1); it makes sure that a disk percentage is always free when moving replicas
+```
+
+In order to use the `remove_disks` endpoint in the Strimzi cluster operator, it would be added to the [`CruiseControlApi`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApi.java) interface and developing the corresponding implementation.
+
+### Implementation
+
+To implement this feature, we will be adding a new mode to the `KafkaRebalanceMode` class.
+* `remove-disks`: It moves replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+You can use this mode by changing the `spec.mode` to `remove-disks` in the `KafkaRebalance` resource.
+
+A `KafkaRebalance` custom resource would look like this.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: remove-disks
+  brokersAndVolumes:
+    - brokerId: 0
+      volumeIds: [1, 2]
```

Comment:
In conjunction with the `mode: remove-disks` it's not clear whether this should be interpreted as 
* "I want broker 0 to end up with just volumes 1 and 2; remove any others". Resources in Kubernetes are normally declarative, so this its a reasonable interpretation for someone who doesn't know the semantics of the resource.
* "I want to remove volumes 1 and 2 from broker 0". `KafkaRebalance` is intrinsically imperative/specifying an action, so this its a reasonable interpretation for someone who know that.

And actually both those interpretations are wrong! The 2nd is closer, but it won't actually remove the volumes (in the kube sense), it just moves the data.

Perhaps renaming `brokersAndVolumes` to something like `moveReplicasOffVolumes` (or `movingReplicasOffVolumes` or some alternative) would make that clearer. I'm not sure we need to include the word broker or even volume, since the items in the array make clear we're talking about volumes within brokers.

Reply:
I don't think this is completely wrong. But I think we also need to keep in mind the existing parts of the API that already exist in this way for a long time. So I wonder if we should prefer inconsistency over something what might not be completely clear. I personally would probably prefer consistency.

Reply:
I agree consistency is a good thing. But good APIs are also a good thing. I wonder if we should take a moment to consider what other CC features we might want to integrate into this CR and how well they they'd fit with `brokersAndVolumes`. Otherwise I can see that the consistency argument might be used to justify an even more distasteful compromise in the future.

Reply:
Okay, I will change the name to `moveReplicasOffVolumes`, that sound good to me